subprojects/core/src/main/java/org/gradle/api/reflect/TypeOf.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.reflect;
+
+import com.google.common.base.Objects;
+import com.google.common.reflect.TypeToken;
```

Comment:
I thought a little more about this.

What we really need here is a way to destructure complex types into its constituents. We could that with a visitor without introducing any new types (other than `Visitor`).

```java
abstract class TypeOf<T> { 
...
  void accept(Visitor v) { ... }

  interface Visitor {
    void visitParameterizedType(TypeOf<?> rawType, List<TypeOf<?>> typeArguments);
    void visitArrayType(TypeOf<?> componentType);
    ...
}
```

What do you think, @adammurdoch?